[Modernization] Replace WTF::makeReversedRange with std::views::reverse

weinig opened the pull request:
#### 3f05dc5ec966a1f1d630d0b6130cc0cb988a5210
<pre>
[Modernization] Replace WTF::makeReversedRange with std::views::reverse
<a href="https://bugs.webkit.org/show_bug.cgi?id=301825">https://bugs.webkit.org/show_bug.cgi?id=301825</a>

Reviewed by Darin Adler.

Replaces uses of `WTF::makeReversedRange` with `std::views::reverse`.

To allow `std::views::reverse` to work with `WTF::ListHashSet`, the
`WTF::ListHashSet` iterators needed to be updated to conform to the
`std::bidirectional_iterator` concept, which required adding postfix
decrement operators that had been previously omitted intentionally.
The goal of the intentional omission was to avoid accidental performance
issues when manually using iterators, but we mostly use try to avoid
direct iterator usage and being able to use std::ranges is a good tradeoff.

* Source/WTF/wtf/CryptographicallyRandomNumber.cpp:
* Source/WTF/wtf/IteratorRange.h:
* Source/WTF/wtf/ListHashSet.h:
* Source/WTF/wtf/URLHelpers.cpp:
* Source/WebCore/accessibility/AXSearchManager.cpp:
* Source/WebCore/accessibility/AccessibilityObject.cpp:
* Source/WebCore/animation/DocumentTimeline.cpp:
* Source/WebCore/dom/Document.cpp:
* Source/WebCore/dom/DocumentFullscreen.cpp:
* Source/WebCore/dom/Element.cpp:
* Source/WebCore/editing/markup.cpp:
* Source/WebCore/html/closewatcher/CloseWatcherManager.cpp:
* Source/WebCore/layout/floats/FloatingContext.cpp:
* Source/WebCore/layout/formattingContexts/inline/InlineFormattingUtils.cpp:
* Source/WebCore/layout/formattingContexts/inline/InlineLine.cpp:
* Source/WebCore/layout/formattingContexts/inline/InlineLine.h:
* Source/WebCore/layout/formattingContexts/inline/InlineLineBuilder.cpp:
* Source/WebCore/layout/formattingContexts/inline/display/InlineDisplayContentBuilder.cpp:
* Source/WebCore/layout/formattingContexts/inline/ruby/RubyFormattingContext.cpp:
* Source/WebCore/layout/integration/inline/LayoutIntegrationLineLayout.cpp:
* Source/WebCore/loader/FrameLoader.cpp:
* Source/WebCore/page/EventHandler.cpp:
* Source/WebCore/page/PointerCaptureController.cpp:
* Source/WebCore/page/scrolling/coordinated/ScrollingTreeCoordinated.cpp:
* Source/WebCore/platform/graphics/FontCascade.cpp:
* Source/WebCore/platform/graphics/TransparencyLayerContextSwitcher.cpp:
* Source/WebCore/platform/graphics/ca/GraphicsLayerCA.cpp:
* Source/WebCore/platform/ios/wak/WKView.mm:
* Source/WebCore/rendering/LayerAncestorClippingStack.cpp:
* Source/WebCore/rendering/RenderBlockFlow.cpp:
* Source/WebCore/rendering/TextBoxPainter.cpp:
* Source/WebCore/rendering/svg/SVGTextChunk.cpp:
* Source/WebCore/style/AnchorPositionEvaluator.cpp:
* Source/WebCore/style/ContainerQueryEvaluator.cpp:
* Source/WebCore/style/StyleInvalidator.cpp:
* Source/WebCore/style/StyleScopeRuleSets.cpp:
* Source/WebCore/style/StyleTreeResolver.cpp:
* Source/WebCore/style/Styleable.cpp:
* Source/WebCore/style/values/backgrounds/StyleFillLayers.h:
* Source/WebCore/style/values/borders/StyleBoxShadow.cpp:
* Source/WebCore/style/values/borders/StyleShadowInterpolation.h:
* Source/WebCore/style/values/text-decoration/StyleTextShadow.cpp:
* Source/WebKit/UIProcess/RemoteLayerTree/ios/RemoteLayerTreeViews.mm:
* Source/WebKit/WebProcess/FullScreen/WebFullScreenManager.cpp:
* Tools/TestWebKitAPI/Tests/WTF/IteratorRange.cpp:

Canonical link: <a href="https://commits.webkit.org/302487@main">https://commits.webkit.org/302487@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/347e2568c9a2881a93b8026ac83924eb24efa804

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [❌ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/129265 "1 style error") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/159/builds/1521 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/40101 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/136640 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/80656 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/131136 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/155/builds/1454 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/156/builds/1398 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/98444 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/60/builds/66343 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/132212 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/162/builds/1140 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/18/builds/115794 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [  ~~🧪 api-wpe~~](https://ews-build.webkit.org/#/builders/41/builds/79095 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/154/builds/1066 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/135/builds/33913 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/79919 "Built successfully") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/121256 "Built successfully and passed tests") | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/109519 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/34408 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/139114 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/127717 "Built successfully and passed tests") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/153/builds/1312 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/161/builds/1265 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/106973 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/160/builds/1361 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/112130 "Passed tests") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/21/builds/106812 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/27187 "Built successfully and passed tests") | [  ~~🧪 vision-wk2~~](https://ews-build.webkit.org/#/builders/164/builds/1086 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-intel-wk2~~](https://ews-build.webkit.org/#/builders/137/builds/30649 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/53935 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/158/builds/1384 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/64740 "Built successfully") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/160731 "Built successfully") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/157/builds/1207 "Built successfully") | | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/40112 "Passed tests") | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/163/builds/1243 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/152/builds/1306 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

darinadler commented on Source/WTF/wtf/ListHashSet.h:
```diff
@@ -255,8 +255,12 @@ template<typename ValueArg, typename HashArg> class ListHashSetIterator {
+    {
+        iterator temp = *this;
+        --(*this);
+        return temp;
```
```suggestion
        iterator result = *this;
        --*this;
        return result;
```

darinadler commented on Source/WTF/wtf/ListHashSet.h:
```diff
@@ -350,7 +354,12 @@ template<typename ValueArg, typename HashArg> class ListHashSetConstIterator {
+    {
+        const_iterator temp = *this;
+        --(*this);
+        return temp;
```
```suggestion
        const_iterator result = *this;
        --*this;
        return result;
```

darinadler commented on Source/WebCore/platform/graphics/TransparencyLayerContextSwitcher.cpp:
```diff
@@ -69,7 +70,7 @@ void TransparencyLayerContextSwitcher::beginDrawSourceImage(GraphicsContext& des
 void TransparencyLayerContextSwitcher::endDrawSourceImage(GraphicsContext& destinationContext, const DestinationColorSpace&)
 {
-    for ([[maybe_unused]] auto& filterStyle : makeReversedRange(m_filterStyles)) {
+    for ([[maybe_unused]] auto& filterStyle : m_filterStyles | std::views::reverse) {
```
Does not seem like we need to reverse it since we are using this only for a count.

weinig committed:
[Modernization] Replace WTF::makeReversedRange with std::views::reverse
https://bugs.webkit.org/show_bug.cgi?id=301825

Reviewed by Darin Adler.

Replaces uses of `WTF::makeReversedRange` with `std::views::reverse`.

To allow `std::views::reverse` to work with `WTF::ListHashSet`, the
`WTF::ListHashSet` iterators needed to be updated to conform to the
`std::bidirectional_iterator` concept, which required adding postfix
decrement operators that had been previously omitted intentionally.
The goal of the intentional omission was to avoid accidental performance
issues when manually using iterators, but we mostly use try to avoid
direct iterator usage and being able to use std::ranges is a good tradeoff.

* Source/WTF/wtf/CryptographicallyRandomNumber.cpp:
* Source/WTF/wtf/IteratorRange.h:
* Source/WTF/wtf/ListHashSet.h:
* Source/WTF/wtf/URLHelpers.cpp:
* Source/WebCore/accessibility/AXSearchManager.cpp:
* Source/WebCore/accessibility/AccessibilityObject.cpp:
* Source/WebCore/animation/DocumentTimeline.cpp:
* Source/WebCore/dom/Document.cpp:
* Source/WebCore/dom/DocumentFullscreen.cpp:
* Source/WebCore/dom/Element.cpp:
* Source/WebCore/editing/markup.cpp:
* Source/WebCore/html/closewatcher/CloseWatcherManager.cpp:
* Source/WebCore/layout/floats/FloatingContext.cpp:
* Source/WebCore/layout/formattingContexts/inline/InlineFormattingUtils.cpp:
* Source/WebCore/layout/formattingContexts/inline/InlineLine.cpp:
* Source/WebCore/layout/formattingContexts/inline/InlineLine.h:
* Source/WebCore/layout/formattingContexts/inline/InlineLineBuilder.cpp:
* Source/WebCore/layout/formattingContexts/inline/display/InlineDisplayContentBuilder.cpp:
* Source/WebCore/layout/formattingContexts/inline/ruby/RubyFormattingContext.cpp:
* Source/WebCore/layout/integration/inline/LayoutIntegrationLineLayout.cpp:
* Source/WebCore/loader/FrameLoader.cpp:
* Source/WebCore/page/EventHandler.cpp:
* Source/WebCore/page/PointerCaptureController.cpp:
* Source/WebCore/page/scrolling/coordinated/ScrollingTreeCoordinated.cpp:
* Source/WebCore/platform/graphics/FontCascade.cpp:
* Source/WebCore/platform/graphics/TransparencyLayerContextSwitcher.cpp:
* Source/WebCore/platform/graphics/ca/GraphicsLayerCA.cpp:
* Source/WebCore/platform/ios/wak/WKView.mm:
* Source/WebCore/rendering/LayerAncestorClippingStack.cpp:
* Source/WebCore/rendering/RenderBlockFlow.cpp:
* Source/WebCore/rendering/TextBoxPainter.cpp:
* Source/WebCore/rendering/svg/SVGTextChunk.cpp:
* Source/WebCore/style/AnchorPositionEvaluator.cpp:
* Source/WebCore/style/ContainerQueryEvaluator.cpp:
* Source/WebCore/style/StyleInvalidator.cpp:
* Source/WebCore/style/StyleScopeRuleSets.cpp:
* Source/WebCore/style/StyleTreeResolver.cpp:
* Source/WebCore/style/Styleable.cpp:
* Source/WebCore/style/values/backgrounds/StyleFillLayers.h:
* Source/WebCore/style/values/borders/StyleBoxShadow.cpp:
* Source/WebCore/style/values/borders/StyleShadowInterpolation.h:
* Source/WebCore/style/values/text-decoration/StyleTextShadow.cpp:
* Source/WebKit/UIProcess/RemoteLayerTree/ios/RemoteLayerTreeViews.mm:
* Source/WebKit/WebProcess/FullScreen/WebFullScreenManager.cpp:
* Tools/TestWebKitAPI/Tests/WTF/IteratorRange.cpp:

Canonical link: https://commits.webkit.org/302487@main
